### ISSUE TITLE
Discussion on Pre-trained Model Evaluation and Baseline Details

### ISSUE THREAD

**Issue by Jimmy-Bots** [Miche11eU/PET-Image-Denoising-Using-3D-Diffusion-Model#3]
Dear Authors,

I'm currently working with your project and have been utilizing the pre-trained model you've generously provided.

I've been evaluating its performance on the Ultra-low Dose PET Imaging Challenge dataset. While using common image quality metrics like SSIM and PSNR, I seem to be observing results where the performance on my test setup does not clearly exceed that of a relatively simple 3D U-Net baseline I've implemented.

This outcome felt unexpected, especially considering the potential advancements your model represents. To help me understand this discrepancy better and ensure that my evaluation methodology is as consistent as possible with yours for a fair comparison, I would be very grateful if you could potentially share more information on the following:

Metric Calculation Details: 
Could you please provide details or perhaps even the code snippet used for calculating the quantitative metrics, particularly SSIM and PSNR? I was wondering if there might be specific implementation details that could significantly influence the scores and might differ between my setup and yours.

Baseline Model Specifics: 
I would be very grateful if you could share more detailed information specifically about the baseline model(s) that were discussed or compared against in your paper. Understanding their specific architecture, training configuration (e.g., optimizer, learning rate schedule, loss function), and any specific data loading or augmentation strategies employed during their training would be extremely helpful in accurately setting up a comparable baseline and better understanding the reported results.

Thank you very much for your significant contribution through your paper and project. I deeply appreciate your time and insights looking into this.

Best regards,
Jimmy

**Comment by Jimmy-Bots** [comment]
Upon performing some evaluations based on the project, I utilized a strategy for inference, which about overlapping. In my implementation, to handle these overlapping sections and reconstruct the full volume prediction, I adopted an averaging method. I noted that the article/code didn't specify this particular step, so I chose this common approach. And when developing and training a U-Net baseline, I made sure the data processing aligned with the methods mentioned in the article.
<img width="615" height="341" alt="Image" src="https://github.com/user-attachments/assets/6ede92ee-2426-4339-868d-f44efbfdd7a1" />
<img width="777" height="1245" alt="Image" src="https://github.com/user-attachments/assets/3b75057e-f7e4-403d-8b07-bd94a327b6ae" />

**Comment by zgongkuang** [comment]
Hi @Jimmy-Bots , thanks for testing the model. After looking at your attached image, I feel that the 3D DDPM result is much worse than what we have observed and presented in the EJNMMI paper. Also, somehow I feel that your 3D Unet is too oversmooth. Several questions to you: (1) how do you implement the averaging? how many slices were used in averaging? (2) what's your image scale (I mean, do you use the same scaling as the released model) when you used it as the network input? @Miche11eU  can add more comments on this.

**Comment by Jimmy-Bots** [comment]
Thank you for your quick response and valuable questions.

Regarding the questions you raised:

1.  **Averaging method:** As stated in the paper and reflected in the code, 10 slices were utilized for overlapping between adjacent patches along the axial (Z) dimension. My implementation performed averaging by calculating the mean prediction value for each voxel within this 10-slice overlapping region, combining the outputs from the two corresponding patches.

2.  **Image scale/input:** I used the same Ultra-low Dose PET Imaging Challenge dataset. The original image spacing of 1.65 × 1.65 × 1.65 mm³ was strictly maintained throughout the processing and was not rescaled. In adherence to the procedure described in the paper and the expected input size indicated in the code, I cropped the volumetric inputs to the dimensions of (192 × 288 × 520) before feeding them into the network. The patch size used was indeed (192 × 288 × 96), consistent with the model's input. Therefore, both the input dimensions and the spatial spacing are consistent with the setup described for the released model.


I hope this information clarifies the specifics of my setup and processing steps.

Thanks again for your help!

**Comment by Jimmy-Bots** [comment]
Hi Boxiao @Miche11eU  ,

Thank you so much for your prompt and incredibly helpful response, and for providing the additional details and updated scripts! Your continuous support is greatly appreciated.

I've taken some time to go through the information you provided and the updated code. Here are my thoughts and a couple of questions that came up:

1.  **Preprocessing Pipeline Details:**
    You asked about my preprocessing pipeline. Before inference, I had indeed converted input data to **SUV units** and then **normalized them to the [0, 1] range**. Previously, based on my understanding from the existing code and discussions, I would **truncate the SUV values to a maximum of 4** before normalization. So, in terms of general steps, it seems I was largely aligned.

    However, I noticed from your updated `test.py` that the truncation limit has been adjusted **from 4 to 16**. This is a crucial detail, and I will promptly update my preprocessing to align with this new threshold. I will also proceed with verifying performance specifically on **1/20 low-dose inputs** as you suggested, and will share the results once available.

2.  **Question about Overlapping Patch Code in `scripts/test.py`:**
    Regarding the updated `scripts/test.py` and the new logic for overlapping patches and weighted averaging, specifically this snippet:
    ```python
    # Blend 10-pixel overlapping regions between adjacent patches using weighted averaging
    if i < 4:
        for j in range(0,10):
            quadra_3D_520[:,:, index_b[i]:index_b[i]+10] = quadra_3D_576[:,:, 96*i+86:96*(i+1)]*(1-j/10) + quadra_3D_576[:,:, 96*(i+1):96*(i+1)+10]*(j/10)
    elif i==4:
        for j in range(0,10):
            quadra_3D_520[:,:, index_b[i]:index_b[i]+10] = quadra_3D_576[:,:, 96*i+86:96*(i+1)]*(1-j/10) + quadra_3D_576[:,:, 96*(i+1)+6:96*(i+1)+16]*j/10
    ```
    I have a quick question about the `for j in range(0,10):` loop within this section. It appears that within each iteration of the outer `if` or `elif` block, the target array `quadra_3D_520` is repeatedly assigned within the `j` loop. As a result, it seems that only the final iteration (when `j` is 9) would effectively overwrite the previous assignments for that 10-pixel block. If my understanding is correct, this would mean the weighted average is effectively computed only with `j=9`'s weighting for each voxel, rather than a cumulative blend across all 10 steps. Is it possible there was a slight oversight with the loop's implementation, perhaps intending to sum/average in a different way, or am I misunderstanding its intended functionality?

3.  **Question about `max_value` in `evaluation/get_ssim_psnr.py`:**
    Additionally, in the newly provided evaluation script `evaluation/get_ssim_psnr.py`, I have a question about the `max_value` calculation:
    ```python
    # Load original 3D PET file to get the maximum SUV value
    matfile_data = scio.loadmat(os.path.join(matfile_dir, i + ".mat"))
    pet_SUV = matfile_data['suv']
    max_value = pet_SUV.max()
    ```
    Here, `pet_SUV.max()` is used to determine `max_value`. Could you please clarify if `pet_SUV` refers to: (a) the maximum SUV value across all *original images* (e.g., all low-dose and normal-dose images for that specific subject or dataset, combined), or (b) specifically the maximum SUV value solely from the **normal-dose (ground truth) image** of the *current subject* being evaluated? Understanding this will ensure I apply the normalization consistently for metric calculation.

Thank you once again for your immense help and dedicated support in clarifying these aspects. It's incredibly helpful for aligning my setup and reproducing the excellent results.

Best regards,
Jimmy

**Comment by Jimmy-Bots** [comment]
Hi Boxiao @Miche11eU ,

Thank you so much again for your **incredibly helpful and detailed feedback** in your last response! Your insights were truly **invaluable**.

Following your guidance, I've re-ran my tests, focusing on incorporating the key updates you mentioned, especially the adjustment of the **SUV truncation limit from 4 to 16** and the use of your **updated inference script (`scripts/test.py`)**.

While I haven't yet had the chance to fully delve into the questions I raised about the overlapping code and the `max_value` definition in the evaluation script (due to immediate focus on verifying the core setup), I wanted to quickly share a **very positive preliminary update**.

Due to ongoing hardware resource and time constraints, I was only able to test with **one specific 1/20 low-dose data sample**. However, the results for this sample are now **significantly improved** and **truly align with the excellent performance presented in your paper and my expectations!** I'm **deeply impressed** by the quality of the output.

Here's a quick visualization of the improved output for the tested sample:

| Models | PSNR | SSIM |
| :---------------- | :----- | :------------ |
| 1-20 Low PET | 40.35 | 0.9636 |
| 3D Unet | 43.84 | **0.9918** |
| 3D DDPM | **45.99** | 0.9912 |

<details>
<summary>Visual Comparison</summary>

<img width="898" height="1317" alt="Image" src="https://github.com/user-attachments/assets/5f516a6a-9d2a-49f4-9912-d60875ed0eb7" />
</details>

This successful reproduction is a profound testament to the clarity of your instructions and, more importantly, the **robustness and efficacy of your 3D DDPM model**. It's **genuinely thrilling** to see this level of quality materialized in my local setup.

I want to express my sincere appreciation for your **outstanding work and your paper**. It is undoubtedly a **significant and practical contribution** to the field of medical image reconstruction/denoising, and its real-world applicability is now even clearer to me through this successful experience.

Thank you once again for your **exceptional patience, dedicated support**, and for investing your time to help me align my setup. It's truly rare to receive such dedicated assistance, and I am very grateful.

Best regards,
Jimmy

**Comment by Miche11eU** [comment]
Hi Jimmy @Jimmy-Bots ,

Thank you so much for your incredibly thoughtful follow-up and the detailed feedback! I’m truly glad to hear that the model is now performing as expected in your setup. 

Regarding the normalization process, yes, I previously used a separate preprocessing script where SUV conversion and normalization were performed before inference. To make this clearer, I updated the testing script so that, as long as the input data is already in SUV units and resampled to the correct voxel size, the script should run directly without requiring additional preprocessing.

About the overlapping patch code, you're right, I had mistakenly pasted an earlier incorrect version of that section. It has now been replaced with the correct version. Thank you for catching this.

As for the max_value used in PSNR calculation, I used the maximum SUV value from the normal-dose (ground truth) PET image of each specific subject, which is option (b) you mentioned. 

Once again, it's great to see your updated results. I really appreciate you taking the time to share this. Please feel free to reach out if you have any further questions. 

Best,
Boxiao

**Comment by Dpw506** [comment]
Hi Jimmy @Jimmy-Bots ，
Your exchange was exciting. I noticed that your implementation accurately converts PET format to SUV. Could you please share the code for your SUV conversion process? I would be extremely grateful for your help.
